commons/src/main/scala/za/co/absa/spline/common/rest/RESTClient.scala

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2020 ABSA Group Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package za.co.absa.spline.common.rest
+
+import java.net.URI
+
+import org.apache.commons.io.IOUtils
+import org.apache.http.auth.Credentials
+import org.apache.http.client.methods.{HttpDelete, HttpGet, HttpPost, HttpRequestBase}
+import org.apache.http.entity.StringEntity
+import org.apache.http.impl.auth.BasicScheme
+import org.apache.http.impl.client.HttpClients
+import za.co.absa.commons.lang.ARM
+import za.co.absa.commons.lang.ARM.managed
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class RESTClient(uri: URI, maybeCredentials: Option[Credentials]) {
+  @throws[HttpStatusException]
+  def get(path: String): Future[String] = execHttp {
+    baseUri => new HttpGet(s"$baseUri/$path")
+  }
+
+  @throws[HttpStatusException]
+  def delete(path: String)(implicit ec: ExecutionContext): Future[Unit] = execHttp {
+    baseUri => new HttpDelete(s"$baseUri/$path")
+  }.map(_ => {})
+
+  @throws[HttpStatusException]
+  def post(path: String, body: String)(implicit ec: ExecutionContext): Future[Unit] = execHttp {
+    baseUri =>
+      new HttpPost(s"$baseUri/$path") {
+        setEntity(new StringEntity(body))
+      }
+  }.map(_ => {})
+
+  @throws[HttpStatusException]
+  private def execHttp(method: URI => HttpRequestBase): Future[String] = {
+    val request = {
+      val req = method(uri)
+      maybeCredentials.foreach(credentials => {
+        val authHeader = new BasicScheme().authenticate(credentials, req, null)
+        req.addHeader(authHeader)
+      })
+      req
+    }
+
+    val (respStatusLine, respBody) =
+      for {
+        httpClient <- managed(HttpClients.createDefault)
+        response <- managed(httpClient.execute(request))
+      } yield {
+        val maybeBody = Option(response.getEntity)
+          .map(e => {
+            val encoding = Option(e.getContentEncoding).map(_.getValue).getOrElse("UTF-8")
+            ARM.using(e.getContent) {
+              inputStream =>
+                IOUtils.toString(inputStream, encoding)
+            }
+          })
+        (response.getStatusLine, maybeBody.orNull)
+      }
+
+    respStatusLine.getStatusCode match {
+      case 200 | 201 | 204 =>
+        Future.successful(respBody)
+      case _ =>
+        throw new HttpStatusException(respStatusLine.getStatusCode, s"ArangoDB response: $respStatusLine. $respBody")
```

Comment:
Why not use Future.failed() instead of throwing?

Or opposite question. Is the Future actually useful here? We could return `respBody` immediately.

Reply:
Good point. Actually I'd prefer to keep async nature of the methods. Even though the current implementation uses a blocking Http client, wrapping it with a future makes it easy to change the impl anytime.

Would something like this read better?
```scala
    Future.fromTry(Try(respStatusLine.getStatusCode match {
      case 200 | 201 | 204 =>
        respBody
      case _ =>
        throw new HttpStatusException(respStatusLine.getStatusCode, s"ArangoDB response: $respStatusLine. $respBody")
    }))
```

Reply:
`Future.fromTry(Try` can even be placed on the beginning of the method

Reply:
This would be the easiest to read for me:
```
    respStatusLine.getStatusCode match {
      case 200 | 201 | 204 =>
        Future.successful(respBody)
      case statusCode =>
        Future.failed(new HttpStatusException(statusCode, s"ArangoDB response: $respStatusLine. $respBody"))
    }
```

But wrapping the whole method in `Future.fromTry(Try` would catch all exceptions happening there, but I don't know whether that is something we want.

Both options are ok for me.

Reply:
I'd say if a method returns a `Future` then it shouldn't throw. It often leads to inconsistency and difficulties in error handling.

Reply:
done